include/nbl/video/IGPUCommandPool.h

```diff
@@ -4,38 +4,835 @@
+{
+public:
+    CDrawIndirectCountCmd(core::smart_refctd_ptr<const video::IGPUBuffer>&& buffer, core::smart_refctd_ptr<const video::IGPUBuffer>&& countBuffer)
+        : m_buffer(std::move(buffer)) , m_countBuffer(std::move(countBuffer))
+    {}
+
+private:
+    core::smart_refctd_ptr<const IGPUBuffer> m_buffer;
+    core::smart_refctd_ptr<const IGPUBuffer> m_countBuffer;
+};
+
+class IGPUCommandPool::CDrawIndexedIndirectCountCmd : public IGPUCommandPool::IFixedSizeCommand<CDrawIndexedIndirectCountCmd>
+{
+public:
+    CDrawIndexedIndirectCountCmd(core::smart_refctd_ptr<const video::IGPUBuffer>&& buffer, core::smart_refctd_ptr<const video::IGPUBuffer>&& countBuffer)
+        : m_buffer(std::move(buffer)), m_countBuffer(std::move(countBuffer))
+    {}
+
+private:
+    core::smart_refctd_ptr<const IGPUBuffer> m_buffer;
+    core::smart_refctd_ptr<const IGPUBuffer> m_countBuffer;
+};
+
+class IGPUCommandPool::CBeginRenderPassCmd : public IGPUCommandPool::IFixedSizeCommand<CBeginRenderPassCmd>
+{
+public:
+    CBeginRenderPassCmd(core::smart_refctd_ptr<const video::IGPURenderpass>&& renderpass, core::smart_refctd_ptr<const video::IGPUFramebuffer>&& framebuffer)
+        : m_renderpass(std::move(renderpass)), m_framebuffer(std::move(framebuffer))
+    {}
+
+private:
+    core::smart_refctd_ptr<const video::IGPURenderpass> m_renderpass;
+    core::smart_refctd_ptr<const video::IGPUFramebuffer> m_framebuffer;
+};
+
+class IGPUCommandPool::CPipelineBarrierCmd : public IGPUCommandPool::ICommand
+{
+public:
+    CPipelineBarrierCmd(const uint32_t bufferCount, const core::smart_refctd_ptr<const IGPUBuffer>* buffers, const uint32_t imageCount, const core::smart_refctd_ptr<const IGPUImage>* images)
+        : ICommand(calc_size(bufferCount, buffers, imageCount, images)), m_resourceCount(bufferCount+imageCount)
+    {
+        m_barrierResources = reinterpret_cast<core::smart_refctd_ptr<const core::IReferenceCounted>*>(reinterpret_cast<uint8_t*>(this) + sizeof(CPipelineBarrierCmd));
+        std::uninitialized_default_construct_n(m_barrierResources, m_resourceCount);
+
+        uint32_t k = 0;
+
+        for (auto i = 0; i < bufferCount; ++i)
+            m_barrierResources[k++] = buffers[i];
+
+        for (auto i = 0; i < imageCount; ++i)
+            m_barrierResources[k++] = images[i];
+    }
+
+    ~CPipelineBarrierCmd()
+    {
+        for (auto i = 0; i < m_resourceCount; ++i)
+            m_barrierResources[i].~smart_refctd_ptr();
+    }
+
+    static uint32_t calc_size(const uint32_t bufferCount, const core::smart_refctd_ptr<const IGPUBuffer>* buffers, const uint32_t imageCount, const core::smart_refctd_ptr<const IGPUImage>* images)
+    {
+        return core::alignUp(sizeof(CPipelineBarrierCmd) + (bufferCount+imageCount)*sizeof(core::smart_refctd_ptr<const core::IReferenceCounted>), alignof(CPipelineBarrierCmd));
+    }
+
+private:
+    const uint32_t m_resourceCount;
+    core::smart_refctd_ptr<const core::IReferenceCounted>* m_barrierResources;
```

Comment:
cool trick.... don't declare `m_barrierResources` at all.

Just intantiate it on the fly
```cpp
core::smart_refctd_ptr<const core::IReferenceCounted>* getBarrierResources()
{
   return reinterpret_cast<core::smart_refctd_ptr<const core::IReferenceCounted>*>(this+1);
}
```

Reply:
Resolved.